### PR TITLE
Add combinators for Timeout Buffered writer stream

### PR DESCRIPTION
**Pull request by rnjtranjan**
With Timeout combinators 45,000,000  Bytes written to socket and then drained out in **142.4000** seconds vs **122.0099** secs without Timeout combinators.

### CODE REVIEW

File: src/Streamly/Internal/Network/Socket.hs [hunk 1]
```diff
@@ -497,6 +498,17 @@ writeStrings encode h =
 fromBytesWithBufferOf :: MonadIO m => Int -> Socket -> SerialT m Word8 -> m ()
 fromBytesWithBufferOf n h m = fromChunks h $ AS.arraysOf n m
 
+-- | Transform a stream of @a@ to @(Just a)@ then intersperse a Nothing into the input 
+-- stream after every n elements then interject Nothing every @t@ seconds.
+-- @since 0.8.0
+{-# INLINE chunksOfTimeout #-}
+chunksOfTimeout :: (MonadIO m , IsStream t, MonadAsync m)
+    => Int -> Double -> t m a -> t m (Maybe a)
+chunksOfTimeout n t =
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Network/Socket.hs:507]: 1. This should go in the same module as `chunksOf`/`intervalsOf`. 
2. We could have a corresponding `chunksOf` as well (i.e. without a timeout), but we already have a `chunksOf`. So should we call these `justsOf` and `justsOfTimeout` to indicate contiguous chunks of `Just` values?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
We also need corresponding APIs in the Handle/File modules, that could be a separate PR though. Handle module has benchmarks as well so we can see how it performs.

**Comment by rnjtranjan** [comment]
a PR https://github.com/composewell/streamly/pull/1236 for FileHandle has been created.

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 2]
```diff
@@ -1157,3 +1160,56 @@ toStreamD_ size Array{..} =
...
+        when (elemCount < 0) $ error "toArrayMinChunk: elemCount is negative"
+        liftIO $ newArrayAligned alignSize elemCount
+    step arr@(Array start end bound) x | end == bound = do
+        let p = unsafeForeignPtrToPtr start
+            oldSize = end `minusPtr` p
+            newSize = max (oldSize * 2) 1
+        arr1 <- liftIO $ reallocAligned alignSize newSize arr
+        insertElem arr1 x
+    step arr x = insertElem arr x
+    extract = liftIO . shrinkToFit
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1215]: We should not be writing new custom combinators for this job. We can use existing fold/parse combinators.

File: src/Streamly/Internal/Data/Stream/IsStream/Transform.hs [hunk 3]
```diff
@@ -1666,3 +1668,22 @@ applyAsync = (|$)
...
+-- | Transform a stream of @a@ to @(Just a)@ then intersperse a Nothing into the input
+-- stream after every n elements then interject Nothing every @t@ seconds.
+-- @since 0.8.0
+{-# INLINE justsOfTimeout #-}
+justsOfTimeout :: (MonadIO m, IsStream t, MonadAsync m)
+    => Int -> Double -> t m a -> t m (Maybe a)
+justsOfTimeout n t =
+      interjectSuffix t (return Nothing)
+    . intersperseSuffixBySpan n (return Nothing)
+    . justsOf
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Stream/IsStream/Transform.hs:1689]: As mentioned in #456 this has an undesired behavior. The timeout may occur after the span is emitted and the timeout would force eject a smaller span. Ideally we should reset the timer whenever we emit a span.

**Review by harendra-kumar — COMMENTED** [review]
I am going to close this PR and create a new one as a replacement.

**Comment by harendra-kumar** [comment]
Closing this, it should be implemented differently, I will create a new one.

**Pull request closed without merging**